parquet-column/src/test/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,413 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.parquet.column.values.bloom;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.apache.parquet.Preconditions;
+import org.apache.parquet.bytes.*;
+import org.apache.parquet.column.ParquetProperties;
+import org.apache.parquet.io.ParquetEncodingException;
+import org.apache.parquet.io.api.Binary;
+
+/**
+ * Bloom Filter is a compact structure to indicate whether an item is not in set or probably
+ * in set. Bloom class is underlying class of Bloom Filter which stores a bit set represents
+ * elements set, hash strategy and bloom filter algorithm.
+ *
+ * Bloom Filter algorithm is implemented using block Bloom filters from Putze et al.'s "Cache-,
+ * Hash- and Space-Efficient Bloom Filters". The basic idea is to hash the item to a tiny Bloom
+ * Filter which size fit a single cache line or smaller. This implementation sets 8 bits in
+ * each tiny Bloom Filter. Tiny bloom filter are 32 bytes to take advantage of 32-bytes SIMD
+ * instruction.
+ */
+
+public class Bloom {
+  // Hash strategy available for bloom filter.
+  public enum HASH {
+    MURMUR3_X64_128,
+  }
+
+  // Bloom filter algorithm.
+  public enum ALGORITHM {
+    BLOCK,
+  }
+
+  /**
+   * Default false positive probability value use to calculate optimal number of bits
+   * used by bloom filter.
+   */
+  public final double DEFAULT_FPP = 0.01;
+
+  // Bloom filter data header, including number of bytes, hash strategy and algorithm.
+  public static final int HEADER_SIZE = 12;
+
+  // Bytes in a bucket.
+  public static final int BYTES_PER_BUCKET = 32;
+
+  // Hash strategy used in this bloom filter.
+  public final HASH hash;
+
+  // Algorithm applied of this bloom filter.
+  public final ALGORITHM algorithm;
+
+  // The underlying byte array for bloom filter bitset.
+  private byte[] bitset;
+
+  // A integer array buffer of underlying bitset help setting bits.
+  private IntBuffer intBuffer;
+
+  // A cache to column distinct value (hash)
+  private Set<Long> elements;
+
+  /**
+   * Constructor of bloom filter, if numBytes is zero, bloom filter bitset
+   * will be created lazily and the number of bytes will be calculated through
+   * distinct values in cache.
+   * @param numBytes The number of bytes for bloom filter bitset, set to zero can
+   *                 let it calculate number automatically by using default DEFAULT_FPP.
+   * @param hash The hash strategy bloom filter apply.
+   * @param algorithm The algorithm of bloom filter.
+   */
+  public Bloom(int numBytes, HASH hash, ALGORITHM algorithm) {
```

Comment:
I doubt we are going to use this class for other bloom filter variants, so I don't see the need to pass the algorithm in. Same thing with the hash function. Since there is only one, let's avoid passing it explicitly.

Reply:
Changed to private and provide another constructor.